README/List.agda

```diff
@@ -0,0 +1,426 @@
+
+  -- Properties of the various list operations can be found in:
+
+  open import Data.List.Properties
+
+  lem₆ : ∀ n (xs : List ℕ) → take n xs ++ drop n xs ≡ xs
+  lem₆ = take++drop
+
+  lem₇ : ∀ (xs : List ℕ) → reverse (reverse xs) ≡ xs
+  lem₇ = reverse-involutive
+
+  lem₈ : ∀ (xs ys zs : List ℕ) → (xs ++ ys) ++ zs ≡ xs ++ (ys ++ zs)
+  lem₈ = ++-assoc
+
+------------------------------------------------------------------------
+-- 2. Binary relations over lists
+------------------------------------------------------------------------
+
+-- All binary relations over lists are found in the folder
+-- `Data.List.Relation.Binary`.
+
+------------------------------------------------------------------------
+-- Pointwise
+
+module PointwiseExplanation where
+
+  -- One of the most basic ways to form a binary relation between two
+  -- lists of type `List A`, given a binary relation over `A`, is to say
+  -- that two lists are related if the first elements in the lists are
+  -- related, and the second elements are related and the third pair are
+  -- related etc. This "pointwise" lifting of a relation to lists is
+  -- found in:
+
+  open import Data.List.Relation.Binary.Pointwise
+
+  -- The same syntax to construct a list (`[]` & `_∷_`) is used to
+  -- construct proofs for the `Pointwise` relation. For example if you
+  -- want to prove that one list is strictly less than another list:
+
+  lem₁ : Pointwise _<_ (0 ∷ 2 ∷ 1 ∷ []) (1 ∷ 4 ∷ 2 ∷ [])
+  lem₁ = 0<1 ∷ 2<4 ∷ 1<2 ∷ []
+    where
+    0<1 = s≤s z≤n
+    2<4 = s≤s (s≤s (s≤s z≤n))
+    1<2 = s≤s 0<1
+
+  -- Lists that are related by `Pointwise` must be of the same length.
+  -- For example:
+
+  open import Relation.Nullary using (¬_)
+
+  lem₂ : ¬ Pointwise _<_ (0 ∷ 2 ∷ []) (1 ∷ [])
+  lem₂ (0<1 ∷ ())
+
+------------------------------------------------------------------------
+-- Equality
+
+module EqualityExplanation where
+
+  -- There are many different options for what it means for two different
+  -- lists of type `List A` to be "equal". We will initially consider
+  -- notions of equality that require the list elements to be in the
+  -- same order and later discuss other types of equality.
+
+  -- The most basic option in the former case is simply to use
+  -- propositional equality `_≡_` over lists:
+
+  open import Relation.Binary.PropositionalEquality
+    using (_≡_; sym; refl)
+
+  lem₁ : 1 ∷ 2 ∷ 3 ∷ [] ≡ 1 ∷ 2 ∷ 3 ∷ []
+  lem₁ = refl
+
+  -- However propositional equality is only suitable when we want to
+  -- use propositional equality to compare the individual elements.
+  -- Although a contrived example, consider trying to prove the equality
+  -- of two lists of the type `List (ℕ → ℕ)`:
+
+  lem₂ : (λ x → 2 * x + 2) ∷ [] ≡ (λ x → 2 * (x + 1)) ∷ []
+
+  -- In such a case it is impossible to prove the two lists equal with
+  -- refl as the two functions are not propositionally equal. In the
+  -- absence of postulating function extensionality, the most common
```

Comment:
'Postulating function extensionality' - might be worth adding a link to docs somewhere explaining what and why Agra doesn't do this? Here may be the first time a new user comes across the concept.